File: Solution/652.寻找重复的子树/652.寻找重复的子树.java
/*
 * @Descroption: LeetCode 652. 寻找重复的子树
 * @Author: EmoryHuang
 * @Date: 2022-09-05 10:41:30
 * @Method:
 * dfs
 * 将子树进行序列化表示，作为key，并存入哈希表
 * 若出现相同key则认为出现重复子树
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    Map<String, TreeNode> map = new HashMap<String, TreeNode>();
    Set<TreeNode> ans = new HashSet<TreeNode>();

    public String dfs(TreeNode node) {
        if (node == null)
            return "";
        StringBuilder sb = new StringBuilder();
        sb.append(node.val).append('(');
        sb.append(dfs(node.left)).append(")(");
        sb.append(dfs(node.right)).append(')');
        String key = sb.toString();
        if (map.containsKey(key)) {
            ans.add(map.get(key));
        } else {
            map.put(key, node);
        }
        return key;

    }

    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        dfs(root);
        return new ArrayList<TreeNode>(ans);
    }
}

File: Solution/652.寻找重复的子树/652.寻找重复的子树.py
'''
Descroption: LeetCode 652. 寻找重复的子树
Author: EmoryHuang
Date: 2022-09-05 10:28:30
Method:
dfs
将子树进行序列化表示，作为key，并存入哈希表
若出现相同key则认为出现重复子树
'''


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: Optional[TreeNode]) -> List[Optional[TreeNode]]:
        def dfs(node):
            if not node:
                return ""
            key = f'{node.val}({dfs(node.left)})({dfs(node.right)})'
            subtree = m.get(key, None)
            if subtree:
                ans.add(subtree)
            else:
                m[key] = node
            return key

        m = defaultdict(str)
        ans = set()
        dfs(root)
        return ans